Worlds

demipixel opened the pull request:
I almost cried myself to sleep tonight after I realized I hadn't pulled from master in like 20 commits even though I finished worlds. Luckily you didn't change much :)

- Basics of worlds going. Typing "world" will teleport you to 0, 70, 0 in the nether and generate it if need be.
- A few changes to logging in and setting positions, etc
- serv.worlds, add more worlds by doing serv.worlds.push(new World(...))
- BUG: Time is randomly switching... Is this my fault? Every second it seems to flip back and forth between two "times", meaning it will flash from day to night...? Might be something else, I don't know, because I didn't touch time...
- CODE CRAP: "world" command is stuffed in playerPlugin/world.js. Should I spend the effort of trying to generate a portal both in the overworld (using flint + steel) and in the nether (fully generated by server) or should I simply make a cleaner command until we get around to that?

rom1504 commented:
Pretty good.
A few points I think might improve the code:
1. make serv.worlds an object instead of an array so we can so serv.worlds["overworld"] instead of serv.world[0]
2. instead of doing serv.worlds[player.world] everywhere, what about putting a reference of the world directly in player.world ? (it won't copy the world, js works with references for objects)
3. if player.world is the actual world, make a player.worldName to store the world the player is currently in. (maybe the name of the world should be directly in prismarine-world though, but we can change that later)

Well I'm not entirely sure about 1 and 3. I think 2 would improve things though.

demipixel commented:
Okay, when I do a rebase, it works fine, but then it says "1 behind 19 ahead" or whatever. After the rebase, I had to pull and then merge with origin/worlds or whatever it was, otherwise it wouldn't let me push (obviously).

My biggest issue with your first point is that it's similar to the plugin problem: I'm worried about assuming no two plugins are going to try and take the same name (even though I know serv.worlds.overworld would be cleaner to work with). Sure, we could have a serv.createWorld to do this for us, but that means if two plugins have never met before and use the same world name, ONE of them will crash.

For your second option: I was thinking about that. As long as it's easy to use or compare, I think it would be fine (compare as in playerA.world == playerB.world). I don't know if we'd actually need three if we did this one.

rom1504 commented:
"it wouldn't let me push " : do git push --force (after a rebase)

Yeah I guess we could compare that way.

rom1504 commented:
Ok I just tested it.
That nether is fun.

Introduced bug: can't see the other players. Fix it ;)

As a general rule, when you change stuff in flying-squid, think about testing it with 2 clients, to make sure multiplayer stuff still work ;)

rom1504 commented on src/lib/playerPlugins/world.js:
```diff
@@ -0,0 +1,124 @@
+    return t;
+  }
+
+	function sendChunksAroundPlayer(view, world)
```
please fix the indentation there ;)

rom1504 commented on src/lib/playerPlugins/updatePositions.js:
```diff
@@ -82,4 +82,23 @@ function inject(serv,player)
     player.emit("positionChanged");
   }
+
+  function sendPosition(pos, opt) {
```
what is this opt for ?

rom1504 commented on src/lib/playerPlugins/login.js:
```diff
@@ -243,8 +145,6 @@ function inject(serv,player)
     setGameMode(player.gameMode);
     fillTabList();
-    spawnOthers();
-    spawn();
```
pretty sure you can't do that and the order of stuff need to be respected.

rom1504 commented on src/lib/worldGenerations/nether.js:
```diff
@@ -0,0 +1,31 @@
+var Vec3 = require('vec3');
+
+function generation({level=50}={}) {
+	function generateChunk(chunkX, chunkZ) {
```
same remark here: fix the indentation

rom1504 commented on src/lib/playerPlugins/login.js:
```diff
@@ -40,78 +40,30 @@ function inject(serv,player)
+        player.sendChunksAroundPlayer(player.view).then(() => player.sendingChunks=false);
       }
     });
+    player.on("teleport", function() {
```
No, that's not the way to do it. When a player teleport, its position should be changed (see updatePositions.js), which will then emit positionChanged, and send him the right chunks.

rom1504 commented on src/lib/playerPlugins/world.js:
```diff
@@ -0,0 +1,124 @@
+module.exports = inject;
+
+function inject(serv, player) {
+	function spawn() {
```
that's wrong.

demipixel commented:
> 1) please fix the indentation there ;)

k

> 2) Ah and in the rest of the file, we do 2 spaces indentation, configure your editor to do that.

It is configured... o_O I'm using sublime... Sometimes it changes by file (e.g. if it recognizes a 4-space file)

> 3) what is this opt for ?

It's in the api.md. opt.yaw and opt.pitch

> 4) pretty sure you can't do that and the order of stuff need to be respected.

So you want player.spawn() and the bottom again?

> 5) same remark here: fix the indentation

Ahhhhh, I DID have it on two-space display, but I didn't always have it on "tabs=>spaces". *sigh* Sublime...

> 6) No, that's not the way to do it. When a player teleport, its position should be changed (see updatePositions.js), which will then emit positionChanged, and send him the right chunks.

positionChanged should really be named "moved". The difference I'm making is that "moved" is a relativePositionChange and "teleport" is sent by the server via sendPosition. The reason for doing this is because if we have something like this:

```js
    player.on("positionChanged",function(){
      if(!player.sendingChunks && player.entity.position.distanceTo(player.lastPositionChunkUpdated)>16*32)
      {
        player.sendingChunks=true;
        player.sendChunksAroundPlayer(player.view).then(() => player.sendingChunks=false);
      }
    });
```
If they're at 0,0 in the overworld and then teleport to 0,0 in the nether, it all screws up. (Sure, we could set player.lastPositionChunkUpdated to something else, but we have to assume somebody else might have this problem as well). Differentiating between "I've moved some" VS "I moved to another world or moved 100000 blocks" should be noted.

> 7) that's wrong.

You mean indentation again?

rom1504 commented:
3)
Right, why is there even any parameters at all in sendPosition ? I think it would be better to just set the player.entity.position and player.entity.yaw and player.entity.pitch and then just call sendPosition.

4) I want what was at the bottom to be at the bottom, that new player.spawn is wrong (see 7) (see the difference in https://github.com/mhsjlw/flying-squid/pull/80/files )

6) No, it really should be the same thing for any position in a given world. For a change of world, we might have to adapt that condition yeah, pretty sure it's not needed to make a new event for that.

7) no, look at the previous code, this doesn't do the same thing (named_entity_spawn and all are missing there)

rom1504 commented on src/lib/serverPlugins/world.js:
```diff
@@ -7,14 +7,32 @@ var generations={
+function inject(serv,{generation={"name":"diamond_square","options":{"worldHeight":80}}}={}) {
   generation.options.seed=generation.options.seed || Math.random()*Math.pow(2, 32);
   serv.emit("seed",generation.options.seed);
-  serv.world = new World(generations[generation.name](generation.options),regionFolder);
```
add back the regionFolder please ;)

rom1504 commented on doc/api.md:
```diff
@@ -98,9 +102,9 @@ Array of connected players
 Object uuid to players
 
-#### serv.world
+#### serv.worlds
```
didn't you remove .worlds ?

rom1504 commented on src/lib/worldGenerations/nether.js:
```diff
@@ -0,0 +1,31 @@
+    var chunk=new Chunk();
+    for (var x = 0; x < 16; x++) {
+      for (var z = 0; z < 16; z++) {
+        var bedrockheighttop = 1 + Math.round(Math.random()*3)
```
https://github.com/mhsjlw/flying-squid/issues/82

rom1504 commented on src/lib/serverPlugins/world.js:
```diff
@@ -15,6 +16,23 @@ module.exports = inject;
+  serv.netherworld = new World(generations["nether"]({}));
+  //serv.endworld = new World(generations["end"]({}));
+
+  //serv._worldSync=new WorldSync(serv.worlds[0]);
```
`serv._worldSync=new WorldSync(serv.overworld);` should do it

rom1504 commented on src/lib/playerPlugins/updatePositions.js:
```diff
@@ -82,4 +82,24 @@ function inject(serv,player)
+    if (typeof opt.yaw != 'undefined') player.entity.yaw=opt.yaw;
+    if (typeof opt.pitch != 'undefined') player.entity.pitch=opt.pitch;
+    
+    player._client.write('position', {
```
just keep in mind that this won't move the player after he spawned (got to use the teleport packet for that), maybe the name of the function should reflect that (maybe setInitialPosition)

rom1504 commented:
Working and the code is ok, just need to do the "spawn player if close enough" since we don't send to everybody the position of everybody now.

rom1504 commented:
Finally done :D 
I added commits to fix the remaining issues.
